Clarify fill value representation, change nan representation

jbms opened the pull request:
Fixes https://github.com/zarr-developers/zarr-specs/issues/231.

jbms commented:
This eliminates the special "0x..." and "0b..." fill value syntax for floating point numbers and instead adds a special syntax for specifying the NaN payload consistent with the C99 standard.

However, this does not yet address the "r*" raw types.

jbms commented:
Previously, the `fill_value` attribute was worded to allow the special "0x..." and "0b..." syntax seemingly for any data type. In fact, though, with the current data types, there were only two real use cases for that syntax: specifying a precise NaN representation, and specifying a value for the raw "r*" data types.  Allowing it for all data types creates a conflict though with the future introduction of a "unicode string" or "json value" data type, for which we would want to use a JSON string to indicate the actual value.

Currently, with the change made by this PR, the spec does not indicate how to specify the fill value for a raw "r*" data type.  Presumably for those data types the fill value would indeed need to be specified as a sequence of bytes, either as a hex string, or perhaps a base64 string, or a JSON array of integers 0-255.

It occurred to me, though, that the primary use case listed for the raw "r*" data types is as fallback data types for other fixed-size data types defined by extensions to/future versions of the spec.  The problem with the current fallback data type mechanism is that we don't have a fallback fill value, which leads to the following options:
1. If a fallback data type is specified, the fill value must be valid for both the primary data type and fallback data type. 
 For example, if we define `bfloat16` as a new data type and wanted to support a fallback of `uint16`, then `bfloat16` would need to represent its fill value as an unsigned integer.  Alternatively, if we wanted to support a fallback of `r2`, then `bfloat16` would need to represent its fill value as a raw byte sequence.  For consistency, we could just require all fill values to be specified as raw byte sequences, and eliminate support for the more "natural" JSON representations.
2. If a fallback data type is specified, a fallback fill value must also be specified.
3. When using a fallback data type, any operation that depends on the fill value, such as partially writing a chunk or reading a missing chunk, could just fail.

jbms commented on docs/v3/core/v3.0.rst:
```diff
@@ -597,30 +597,57 @@ mandatory names:
+
+      - A JSON string of the form:
+
+        - ``"inf"``, denoting positive infinity;
```
To be clear, the proposed nan syntax as defined does not force a quiet nan, the number given in the parentheses fully specifies all the mantissa bits.  But indeed it cannot be parsed using strtof/strtod, since those functions use an implementation-defined mapping.

Negative zero can already be specified with regular json syntax, no need for a separate representation for that.

It seemed simpler to just have one way to represent each number so that the implementation does not have to choose between two different representations.

But I suppose if just specifying the full binary representation of the float is preferable, that is fine too.  Note that we would still need to precisely specify which NaN is denoted by "NaN".

rouault replied:
> Negative zero can already be specified with regular json syntax, no need for a separate representation for that.

ah, you're right. My testing was incorrect. I used -0 instead of -0.0 ...

jstriebel replied:
> It seemed simpler to just have one way to represent each number so that the implementation does not have to choose between two different representations.

Good point, having a single representation would be nice. But the special `nan(…)` syntax seems quite custom, and in most cases the simple `nan` would do, which also duplicates `nan(0x1)` then.

> But I suppose if just specifying the full binary representation of the float is preferable, that is fine too. Note that we would still need to precisely specify which NaN is denoted by "NaN".

I assume we will also need a binary representation for raw dtypes, and also for more custom dtypes in the future. Therefore I'd also use this for the specific nan representations. But it would be great if you could still include this as a detail, specifying how to represent a specific nan.

jstriebel commented on docs/v3/core/v3.0.rst:
```diff
@@ -597,30 +597,57 @@ mandatory names:
-    For any data type, the ``fill_value`` is required. The literal
-    ``null`` is not permitted. The fill value needs to be defined
-    so that the data is independent of implementation details. Internally
-    implementations may provide a default ``fill_value``, but that must
-    be converted to a fixed value in the stored metadata.
```
I think it might be useful to keep this sentence, to distinguish between metadata and library API level.

jbms replied:
Done.

jstriebel reviewed (commented):
Thanks a ton @jbms!

jbms commented:
I reverted back to the "+Infinity", "-Infinity", 'NaN" syntax used by zarr v2, and added a hex representation for floats in place of the "nan(0xYYYY)" syntax.  I used the "big endian" representation since that is by far the more common way to display the raw bytes of a float as a hexadecimal integer.

For the "rN" raw data types, I instead used an array of numbers, [0, 255].  But I could change this back to the "0x..." syntax if that is preferred.

rouault commented on docs/v3/core/v3.0.rst:
```diff
@@ -717,7 +750,7 @@ compressed using gzip compression prior to storage::
             }
         }],
-        "fill_value": "NaN",
+        "fill_value": "nan(0x7)",
```
```suggestion
        "fill_value": "NaN",
```

rouault commented on docs/v3/core/v3.0.rst:
```diff
@@ -597,30 +597,63 @@ mandatory names:
+      components respectively, where each component is specified as defined
+      above for floating point number.
+
+      For example, ``[1, 2]`` indicates ``1 + 2i`` and ``["-inf", "nan(0x7)"]``
+      indicates a complex number with real component of -inf and imaginary
+      component of positive NaN with payload of ``0x7``.
```
```suggestion
      For example, ``[1, 2]`` indicates ``1 + 2i`` and ``["-Infinity", "NaN"]``
      indicates a complex number with real component of -inf and imaginary
      component of NaN.
```

rouault commented on docs/v3/core/v3.0.rst:
```diff
@@ -597,30 +597,63 @@ mandatory names:
+
+      - A JSON string of the form:
+
+        - ``"+Infinity"``, denoting positive infinity;
```
Note that Zarr V2 uses "Infinity" without leading '+' . Is the discrepancy ZarrV2 vs ZarrV3 intended ?

jbms replied:
That wasn't intentional --- I changed it back.

jbms commented:
@jstriebel Is there anything left to do?

jstriebel reviewed (approved):
LGTM, thanks a lot @jbms! And sorry for coming back to this so late.
